### ISSUE TITLE
Consider dropping numpy dependency

### ISSUE THREAD

**Issue by layday** [maxbachmann/rapidfuzz#123]
I use RapidFuzz in a GUI which I bundle into a stand-alone app using [briefcase](https://briefcase.readthedocs.io/). After upgrading to v1.6.0, I found that my app's size has increased by about 80 MB (!) on macOS, owing to numpy.  Looking at the commit which introduced the numpy dependency, it doesn't look like it's used for anything other than to generate an array in a user-facing function - would you consider dropping numpy as a dependency?

**Comment by maxbachmann** [comment]
Ouch, I did not think about a use case like this. I am personally not super happy about the numpy dependency either, since it makes compiling rapidfuzz a lot slower (1h 20 min vs 40 min). As you identified I use a numpy matrix as result of `process.cdist`.

I am generally fine with replacing this usage, but I am not sure what would be a better alternative. 

## Alternative 1:
I could use a custom matrix class
+ should be about similar fast to create, since I can directly access the buffer inside similar to the way I did with numpy
- users probably want to operate on the matrix and use all the nice matrix features of numpy. So users would need to copy it to a numpy matrix inside Python which is often slower than the whole string matching.

## Alternative 2:
I might be able to reduce the scope in which numpy is imported. However I have no clue whether this is possible in cython:
```
def cdist():
    import numpy as np
    np....
```

## Alternative 3
I could however move cdist to a different Python module, which I try to import in `process.py`:

```
from rapidfuzz.cpp_process import extract, extractOne, extract_iter
try:
    from rapidfuzz.cpp_cdist import cdist
except ImportError:
    def cdist(queries, choices, *, scorer=ratio, processor=None, score_cutoff=None, **kwargs):
        raise NotImplementedError("implementation requires numpy to be installed")
```

Would briefcase need the dependency to be removed from https://github.com/maxbachmann/RapidFuzz/blob/main/setup.cfg, or is it possible to manually tell it not to bundle a certain dependency? Do you have any other better ideas how to handle this

**Comment by layday** [comment]
briefcase invokes `pip install` with a custom `--prefix`.  I don't think there's a way to instruct pip to omit a particular dependency.  I could theoretically patch briefcase to `pip uninstall numpy` - though it'd be a lot easier for me to simply constrain RapidFuzz to `<1.6` until a solution can be found.

**Comment by maxbachmann** [comment]
After thinking about this a bit more I think the best solution would be to use a [custom array container](https://numpy.org/doc/stable/user/basics.dispatch.html) and only import numpy inside the `__array__` function.
```python
class MyArray:
    ...
    def __array__(self, dtype=None):
        import numpy as np
        ...
```
Since this should only be called by numpy:
```
np.asarray(MyArray())
```
it is probably fine to assume that whoever calls the function has numpy installed and therefor remove it from the requirements.

**Comment by maxbachmann** [comment]
I gave this a bit more thought and I think the following would be the best course of actions:
- keep using a numpy array as a return value for now, since it is the most useful type to work with afterwards
- place all parts using numpy in a separate module and make the dependency optional. This way it is not possible to use cdist without numpy. However it is possible to use the rest of the library without it. 
- In the future I might create a simple matrix type, which is used as a fallback when numpy is not available. 

At least for now I decided against always using a custom array forthe following reason:
- the custom matrix would be very limited in function, so people would need to convert it to a numpy array most of the time
  - When using fast similarities like fuzz.ratio, the copy can reduce the runtime significantly
  - It reduces the maximum of data cdist can work with, since the matrix copy temporarily requires twice the memory
 
I think this should work fine for your use case as long as you do not plan to use `cdist`. Let me know in case you know a better solution to this.
